Calculate throttle based harmonic notch using FFT averaging

andyp1per opened the pull request:
Users struggle with harmonic notch setup. On 2Mb boards we have the possibility of using the in-flight FFT to setup the throttle-based harmonic notch. Normally the in-flight FFT can struggle to track noise peaks in noisy builds, but if the FFT's are averaged over a factor of N then the noise is reduced by 1/sqrt(N). This PR introduces the ability to average FFTs in-flight controlled by RC input (function 163). The user hovers, engages 163, hovers for some period, disengages 163 and the lands. The function will setup the appropriate harmonic notch options and scaling factors such that if the user then reboots they will have an enabled and accurate throttle-based notch.

Setup procedure:

1. Enable the FFT - ```FFT_ENABLE = 1```
2. Reboot
3. Put function 162 on an RC switch, e.g. ```SERVO9_OPTION = 162```
4. Make sure the servo channel is low
5. Take off and hover
6. Set the servo channel to high
7. Hover for 30s or longer
8. Set the channel to low again. The notch will have been configured and engaged at this point so you may see improvements in flight performance. To see more improvements rerun autotune
9. You can repeat this procedure if you are not happy with the results

andyp1per commented:
Done

tridge reviewed (changes requested):
consider allocating when disarmed if the RCn_OPTION exists

andyp1per commented:
@tridge I have addressed your comments and tightened up the testing. Currently it's not possible to enable the notch dynamically because it involves allocating the filters which is done at startup. It would be possible to change this to be dynamic with some work, but probably should be a separate PR and I am still not sure about creating filters in-flight. Another option would be simply to enable the notch by default but keep the reference at 0.  This would allocate the data structures without them being used.

andyp1per commented:
@tridge we could also consider switching on FFT by default on H7

tridge reviewed (commented):
need to find what memory cost this is compared to master

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -921,13 +920,12 @@ AP_InertialSensor::init(uint16_t loop_rate)
     for (uint8_t i=0; i<get_gyro_count(); i++) {
         // only allocate notches for IMUs in use
         if (_use[i]) {
-            if (_harmonic_notch_filter.enabled()) {
```
make it depend on notch enable || fft enable?

andyp1per replied:
Fixed

andyp1per commented:
Tested on my miniquad

andyp1per commented:
Memory on this branch:
```
ThreadsV2
ISR           PRI=255 sp=0x20010000 STACK=1280/1536
ArduCopter    PRI=182 sp=0x20010600 STACK=5312/7168
idle          PRI=  1 sp=0x200236B0 STACK=144/352
UART_RX       PRI= 60 sp=0x2004EFB8 STACK=768/1104
OTG1          PRI= 60 sp=0x2004E6D8 STACK=264/656
monitor       PRI=183 sp=0x20020668 STACK=1016/1360
timer         PRI=181 sp=0x20021A88 STACK=1488/1872
rcout         PRI=181 sp=0x20021178 STACK=432/848
rcin          PRI=177 sp=0x20020BF0 STACK=416/1360
io            PRI= 58 sp=0x2001FCE0 STACK=1496/2384
storage       PRI= 59 sp=0x20021500 STACK=832/1360
UART1         PRI= 60 sp=0x2004DF88 STACK=312/656
UART3         PRI= 60 sp=0x2004D8A8 STACK=272/656
UART6         PRI= 60 sp=0x2004D1C8 STACK=304/656
UART7         PRI= 60 sp=0x2004CAE8 STACK=368/656
I2C0          PRI=176 sp=0x2004B0A8 STACK=904/1360
OSD           PRI= 59 sp=0x20046E38 STACK=920/1616
UART2         PRI= 60 sp=0x20047B58 STACK=248/656
log_io        PRI= 59 sp=0x20047E38 STACK=1504/1912
SPI4          PRI=181 sp=0x20038F48 STACK=720/1360
FTP           PRI= 58 sp=0x200323D0 STACK=1440/2896

MemInfoV1
START=0x20010000 LEN=256k FREE=  2592 LRG=   936 TYPE=0
START=0x20000000 LEN= 64k FREE= 52176 LRG= 52120 TYPE=1
```
